File: Programmers/Level2/타겟넘버/타겟넘버.js
let answer = 0;

function DFS(idx, numbers, value, target) {
    if (idx == numbers.length) {
        if (target == value) {
            answer += 1;
        }
        return;
    }
    DFS(idx + 1, numbers, value + numbers[idx], target);
    DFS(idx + 1, numbers, value - numbers[idx], target);
}

function solution(numbers, target) {
    DFS(0, numbers, 0, target);
    return answer;
}

function solution(numbers, target) {
    let answer = [0];

    for (let n of numbers) {
        let tmp = [];
        for (let i of answer) {
            tmp.push(i + n);
            tmp.push(i - n);
        }
        answer = tmp;
    }
    return answer.filter((e) => e === target).length;
}

File: Programmers/Level2/타겟넘버/타겟넘버.py
# dfs 풀이
answer = 0

def DFS(idx, numbers, value, target):
    global answer

    if idx == len(numbers):
        if target == value:
            answer += 1
        return

    DFS(idx + 1, numbers, value + numbers[idx], target)
    DFS(idx + 1, numbers, value - numbers[idx], target)
    
def solution(numbers, target):
    global answer
    DFS(0, numbers, 0, target)
    return answer

# bfs 풀이
def solution(numbers, target):
    answer = [0]
    
    for n in numbers:
        tmp = []
        for i in answer:
            tmp.append(i + n)
            tmp.append(i - n)
        answer = tmp
        
    return answer.count(target)